### ISSUE TITLE
Routine bioconda build ?

### ISSUE THREAD

**Issue by fubar2** [galaxyproject/planemo#1084]
tl;dr Latest conda build 0.70 is available but doesn't do some useful things. A naive attempt at building a 0.72.0 pypi planemo package in conda
```
conda skeleton pypi planemo
conda build planemo
```
ends in an apparently unsatisfiable dependencies. Is it just me? Conda info below.

Interesting challenge to thread the needle with compromise(s) that won't break everything in case anyone competent wants to take a whack ? 

   return build_tree(sorted(absolute_recipes), config, stats, build_only=build_only, post=post,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/build.py", line 3012, in build_tree
    packages_from_this = build(metadata, stats,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/build.py", line 1982, in build
    output_metas = expand_outputs([(m, need_source_download, need_reparse_in_env)])
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/render.py", line 762, in expand_outputs
    for (output_dict, m) in deepcopy(_m).get_output_metadata_set(permit_unsatisfiable_variants=False):
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/metadata.py", line 2062, in get_output_metadata_set
    conda_packages = finalize_outputs_pass(ref_metadata, conda_packages, pass_no=0,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/metadata.py", line 735, in finalize_outputs_pass
    fm = finalize_metadata(om, parent_metadata=parent_metadata,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/render.py", line 536, in finalize_metadata
    build_unsat, host_unsat = add_upstream_pins(m,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/render.py", line 411, in add_upstream_pins
    host_deps, host_unsat, extra_run_specs_from_host = _read_upstream_pin_files(m, 'host',
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/render.py", line 376, in _read_upstream_pin_files
    deps, actions, unsat = get_env_dependencies(m, env, m.config.variant,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/render.py", line 144, in get_env_dependencies
    actions = environ.get_install_actions(tmpdir, tuple(dependencies), env,
  File "/home/ross/miniconda3/lib/python3.8/site-packages/conda_build/environ.py", line 789, in get_install_actions
    raise DependencyNeedsBuildingError(exc, subdir=subdir)
conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform linux-64: {"markupsafe[version='>=0.23']", "python_abi[version='2.7.*|3.6',build='*_pypy36_pp73|*_cp27mu']", "requests[version='>=2.0.1,<3.0.0|>=2.0.1,<=3.0.0']", "bioblend[version='>=0.10.0']", "certifi[version='>=2016.09|>=2016.9.26']", 'html5lib', 'six', "certifi[version='>=2016.09|>=2016.9.26|>=2017.4.17']", "python_abi[version='2.7.*|3.6|3.6.*|3.7.*|3.8.*',build='*_cp37m|*_cp38|*_cp36m|*_pypy36_pp73|*_cp27mu']", 'python_abi=3.8[build=*_cp38]', "requests[version='>=1.0|>=2.6.1']", "galaxy-util[version='>=19.9|>=20.5']", "_openmp_mutex[version='>=4.5']", "urllib3[version='>=1.21.1,<1.24|>=1.21.1,<1.25|>=1.21.1,<1.26,!=1.25.0,!=1.25.1']", "pypy3.6[version='7.3.*|7.3.0.*|7.3.1.*']", 'pip', 'setuptools', "pypy3.6[version='>=7.3.1']", "gdbm[version='>=1.18,<1.19.0a0']", "requests[version='>=2.20.0|>=2.4.3,!=2.12.0,!=2.12.1|>=2.4.3']", "six[version='>=1.9.0,<2']", "python_abi[version='3.6|3.6.*|3.8.*|3.7.*',build='*_cp37m|*_pypy36_pp73|*_cp36m|*_cp38']", "certifi[version='>=2017.4.17']", 'pyyaml', "pathlib2[version='>=2.3.2']", 'wheel', "python_abi[version='2.7.*|3.7.*|3.6.*|3.6|3.8.*',build='*_cp37m|*_cp27mu|*_cp36m|*_pypy36_pp73|*_cp38']", "galaxy-util[version='>=19.9|>=20.1.0.dev0|>=20.5']", 'bioblend', "configparser[version='>=3.5']", "python_abi[version='2.7.*|3.6.*|3.6|3.8.*|3.7.*',build='*_cp37m|*_cp36m|*_cp38|*_pypy36_pp73|*_cp27mu']", 'jinja2', "html5lib[version='>=0.999|>=0.999,!=0.9999,!=0.99999,<0.99999999|>=0.99999999|>=0.99999999pre,!=1.0b1,!=1.0b2,!=1.0b3,!=1.0b4,!=1.0b5,!=1.0b6,!=1.0b7,!=1.0b8']", "python_abi[version='2.7.*|3.6.*|3.6|3.7.*|3.8.*',build='*_cp37m|*_cp38|*_pypy36_pp73|*_cp36m|*_cp27mu']", "requests[version='>=2.20.0']", "python_abi[version='3.6|3.6.*|3.7.*|3.8.*',build='*_cp37m|*_cp38|*_cp36m|*_pypy36_pp73']", "python_abi[version='2.7.*|3.6.*|3.6|3.8.*|3.7.*',build='*_cp37m|*_cp38|*_pypy36_pp73|*_cp36m|*_cp27mu']", "pypy3.6[version='7.3.*|7.3.0.*|7.3.1.*|>=7.3.1']", 'packaging', 'boltons', 'galaxy-containers', "pypy3.6[version='7.3.0.*|7.3.1.*|>=7.3.1']", 'docutils', 'markupsafe', 'cython', "typing[version='>=3.5.3']", "lxml[version='>=3.3.5']", "python_abi[version='2.7.*|3.8.*|3.6|3.6.*|3.7.*',build='*_cp37m|*_cp38|*_cp27mu|*_pypy36_pp73|*_cp36m']", "urllib3[version='>=1.21.1,<1.22|>=1.21.1,<1.23|>=1.21.1,<1.24|>=1.21.1,<1.25|>=1.21.1,<1.26,!=1.25.0,!=1.25.1']", 'requests', "requests[version='>=2.20.0|>=2.4.3,!=2.12.0,!=2.12.1']", 'pathlib2', "typing[version='>=3.6.2|>=3.7.4']", "six[version='>=1.9.0']", "packaging[version='>=20.2']", 'python_abi=2.7[build=*_cp27mu]', 'python_abi==3.6[build=*_pypy36_pp73]', "galaxy-util[version='>=20.1.0.dev0']", "python_abi[version='3.6|3.6.*|3.8.*|3.7.*',build='*_cp37m|*_cp38|*_cp36m|*_pypy36_pp73']", "pypy3.6[version='7.3.0.*|7.3.1.*']", "six[version='>=1.8.0']", 'ca-certificates', "python_abi[version='3.6.*|3.8.*|3.7.*',build='*_cp37m|*_cp38|*_cp36m']", "galaxy-containers[version='>=19.9|>=20.5']", "python_abi[version='2.7.*|3.6.*|3.8.*|3.7.*|3.6',build='*_cp37m|*_cp38|*_cp36m|*_cp27mu|*_pypy36_pp73']", "galaxy-tool-util[version='>=19.9.1']"}

Conda info for the record:

```
(base) 1 ross@rosspn50:~/galaxythrow$ conda info

     active environment : base
    active env location : /home/ross/miniconda3
            shell level : 1
       user config file : /home/ross/.condarc
 populated config files : /home/ross/.condarc
          conda version : 4.8.5
    conda-build version : 3.20.2
         python version : 3.8.3.final.0
       virtual packages : __glibc=2.31
       base environment : /home/ross/miniconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://conda.anaconda.org/bioconda/linux-64
                          https://conda.anaconda.org/bioconda/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/ross/miniconda3/pkgs
                          /home/ross/.conda/pkgs
       envs directories : /home/ross/miniconda3/envs
                          /home/ross/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.5 requests/2.24.0 CPython/3.8.3 Linux/5.8.0-050800-generic ubuntu/20.04.1 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False


```

**Comment by simonbray** [comment]
It seems there is a PR open here: https://github.com/bioconda/bioconda-recipes/pull/23718

It looks like some python2/3 trouble to me.

**Comment by fubar2** [comment]
Ah - thanks @simonbray.
Building that meta.yaml now. 

It gives a different list - python_abi seems to have gone and there sure seems to be some disagreement about python 2 or 3 or something....

onda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform linux-64: {"python[version='2.7.*|3.5.*|3.6.*|3.4.*']", "python[version='3.7.*|3.8.*']", "gdbm[version='>=1.18,<1.19.0a0']", "python[version='2.7.*|3.5.*|3.6.*|3.4.*|>=3|>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.5,<3.6.0a0']", 'six', "requests[version='>=1.0']", "python[version='>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|>=3.5,<3.6.0a0']", "python[version='2.7.*|3.5.*|3.6.*|>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|3.4.*|>=3.5,<3.6.0a0']", "galaxy-util[version='>=19.9|>=20.5']", 'scandir', "certifi[version='>=2016.09|>=2016.9.26']", "requests[version='>=2.6.1']", "requests[version='>=2.20.0|>=2.4.3,!=2.12.0,!=2.12.1|>=2.4.3']", "python[version='>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0']", "requests[version='>=2.20.0']", "requests[version='>=2.0.1,<3.0.0|>=2.0.1,<=3.0.0']", 'bz2file', "bioblend[version='>=0.10.0']", "requests[version='>=2.12.4|>=2.12.4,<3|>=2.18.4,<3|>=2.5.3|>=2.14.2,<3']", "galaxy-tool-util[version='>=19.9.1']", 'pyyaml', "python[version='2.7.*|3.5.*|3.6.*|>=3|>=2.7,<2.8.0a0|>=3.8,<3.9.0a0|>=3.5|3.4.*|>=3.7,<3.8.0a0|>=3.6,<3.7.0a0|>=3.5,<3.6.0a0']", "python[version='2.7.*|3.5.*|3.6.*|>=3.5,<3.6.0a0|>=3.8,<3.9.0a0|3.4.*|>=3|>=3.5']", "python[version='2.7.*|3.5.*|3.6.*|>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.6|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|>=3.5,<3.6.0a0']", "urllib3[version='>=1.21.1,<1.24|>=1.21.1,<1.25|>=1.21.1,<1.26,!=1.25.0,!=1.25.1']", "python[version='3.7.*|3.8.*|>=3.5']", 'python', "python[version='2.7.*|3.5.*|3.6.*|3.4.*|>=3.7,<3.8.0a0|>=3.6,<3.7.0a0|>=3.8,<3.9.0a0|>=2.7,<2.8.0a0|>=3.5,<3.6.0a0']", "python[version='2.7.*|3.5.*|3.6.*|>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=3.8,<3.9.0a0|>=3.5,<3.6.0a0|3.4.*']", "python[version='>=3.5']", "python[version='2.7.*|3.4.*|3.5.*|3.6.*|>=3.6,<3.7.0a0|>=2.7,<2.8.0a0|>=3.7,<3.8.0a0|>=3.5,<3.6.0a0']", 'docutils', "requests[version='>=2.20.0|>=2.4.3,!=2.12.0,!=2.12.1']", "typing[version='>=3.5.3']", "html5lib[version='>=0.999|>=0.999,!=0.9999,!=0.99999,<0.99999999|>=0.99999999|>=0.99999999pre,!=1.0b1,!=1.0b2,!=1.0b3,!=1.0b4,!=1.0b5,!=1.0b6,!=1.0b7,!=1.0b8']", "certifi[version='>=2016.09|>=2016.9.26|>=2017.4.17']", "python[version='2.7.*|3.5.*|3.6.*|>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|>=3.5,<3.6.0a0|3.4.*']", "_openmp_mutex[version='>=4.5']", "yaml[version='>=0.1.7,<0.2.0a0|>=0.2.2,<0.3.0a0|>=0.2.5,<0.3.0a0']", "lxml[version='>=3.3.5']", 'bioblend', "python[version='2.7.*|3.5.*|3.6.*|3.4.*|>=3.6,<3.7.0a0|>=2.7,<2.8.0a0|>=3.7,<3.8.0a0']", "python[version='2.7.*|3.5.*|3.6.*|>=3|3.4.*|>=3.8,<3.9.0a0|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=2.7,<2.8.0a0|>=3.5,<3.6.0a0']", "urllib3[version='>=1.21.1,<1.22|>=1.21.1,<1.23|>=1.21.1,<1.24|>=1.21.1,<1.25|>=1.21.1,<1.26,!=1.25.0,!=1.25.1']", "python[version='2.7.*|3.5.*|3.6.*|3.4.*|>=3.7,<3.8.0a0|>=3.6,<3.7.0a0|>=2.7,<2.8.0a0|>=3.5,<3.6.0a0']", "python[version='>=3.5,<3.6.0a0|>=3.8,<3.9.0a0']", 'jinja2', "galaxy-containers[version='>=19.9|>=20.5']", "six[version='>=1.9.0,<2']", 'typing', "six[version='>=1.9.0']", "python[version='>=2.7,<2.8.0a0|>=3.5|>=3.7,<3.8.0a0|>=3.6,<3.7.0a0']", "galaxy-util[version='>=19.9|>=20.1.0.dev0|>=20.5']", "python[version='>=3.8,<3.9.0a0']", "six[version='>=1.8.0']", "python[version='2.7.*|3.5.*|3.6.*|3.8.*|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0|>=2.7,<2.8.0a0|>=3.5,<3.6.0a0|3.4.*|>=3.4,<3.5.0a0']", 'markupsafe', "typing[version='>=3.6.2|>=3.7.4']", "python[version='>=3.5|>=3.8,<3.9.0a0']", "certifi[version='>=2017.4.17']", "markupsafe[version='>=0.23']", 'html5lib', "python[version='2.7.*|3.5.*|3.6.*|3.4.*|>=3.5,<3.6.0a0|>=3.8,<3.9.0a0|3.7.*']", "python[version='2.7.*|3.5.*|3.6.*|>=2.7,<2.8.0a0|>=3.6,<3.7.0a0|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|>=3.5,<3.6.0a0']", "python[version='3.4.*|3.8.*|3.7.*']", 'setuptools', "setuptools[version='>=31.0.1']", "python[version='2.7.*|3.5.*|3.6.*|>=2.7,<2.8.0a0|>=3.5,<3.6.0a0|>=3.7,<3.8.0a0|>=3.8,<3.9.0a0|>=3.6,<3.7.0a0|3.4.*']", "python[version='>=3']", 'requests'}

**Comment by fubar2** [comment]
So, to fix the python 2/3 fight, I tried
`conda skeleton pypi --recursive planemo`

Went well until:

```

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Applying patch: '/tmp/tmp08wyb4s5conda_skeleton_docutils-0.16.tar.gz/pypi-distutils.patch'
patching file core.py
Hunk #1 succeeded at 167 with fuzz 2 (offset 1 line).

Leaving build/test directories:
  Work:
 /home/ross/miniconda3/conda-bld/skeleton_1600857314936/work 
  Test:
 /home/ross/miniconda3/conda-bld/skeleton_1600857314936/test_tmp 
Leaving build/test environments:
  Test:
source activate  /home/ross/miniconda3/conda-bld/skeleton_1600857314936/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol 
  Build:
source activate  /home/ross/miniconda3/conda-bld/skeleton_1600857314936/_build_env 


Request to fetch https://pypi.io/pypi/bz2file;/json failed with status: 404

```
at which point I think I might be better off back under the doona :)

**Comment by simonbray** [comment]
The newest versions of ephemeris (0.10.3+) are strangely only available for python2. I've tried rebuilding the ephemeris conda package, hopefully that will solve it.

No idea about your latest error. :)

**Comment by fubar2** [comment]
I got it to build with --no-test 

In the conda build log, I learn that my ubuntu exhibited "Windows file locking mis-bahaviour". Should I call an exorcist?

INFO conda_build.utils:shutil_move_more_retrying(1988): shutil.move(work)=/home/ross/miniconda3/conda-bld/planemo_1600858838411/work, dest=/home/ross/miniconda3/conda-bld/planemo_1600858838411/work_moved_planemo-0.72.0-py_0_linux-64_main_build_loop)
Failed to rename work directory despite sleeping and retrying. This is some Windows file locking mis-bahaviour.
ERROR:conda_build.utils:Failed to rename work directory despite sleeping and retrying. This is some Windows file locking mis-bahaviour.
ERROR conda_build.utils:shutil_move_more_retrying(2003): Failed to rename work directory despite sleeping and retrying. This is some Windows file locking mis-bahaviour.

**Comment by fubar2** [comment]
I think you're right, but a conspiracy theory about sub-aether back channel leakage from the windows subsystem for linux would not seem out of the ordinary on the interwebs in these strange days. :)

**Comment by simonbray** [comment]
I forgot to report back here, but the new conda package got merged last week fwiw

**Comment by fubar2** [comment]
Yes - thanks @simonbray - closing this now.
